### DISCUSSION TITLE
How do I use web::block in a websocket application?

### DISCUSSION THREAD

**Discussion by dessalines** [actix/actix-web#1455] [Q&A] [answered]
My issue: all websocket requests in the example websocket_chat example, are blocking. This means that when I pass an r2d2 connection pool into my ChatServer actor, the query blocks all threads.

Does anyone have `web::block` working in a websocket application? 

I have a websocket app based very closely on the websocket_chat example, but I'm finding the thread is blocked when doing multiple actions that require using an r2d2 connection pool. In the r2d2 examples, it says to use web::block, but there's no examples of this working for a websocket application.

**Comment by dessalines** [comment]
This might be a relevant section, perhaps `wait` is blocking every next action:

```rust
      ws::Message::Text(text) => {
        let m = text.trim().to_owned();
        info!("Message received: {:?} from id: {}", &m, self.id);

        self
          .cs_addr
          .send(StandardMessage {
            id: self.id,
            msg: m,
          })
          .into_actor(self)
          .then(|res, _, ctx| {
            match res {
              Ok(res) => ctx.text(res),
              Err(e) => {
                error!("{}", &e);
              }
            }
            actix::fut::ready(())
          })
          .wait(ctx);
      }
```

Its also been suggested that I use a `SyncArbiter` and `SyncContext` for the main ChatServer actor, but this has been really difficult to implement.

**Comment by dessalines** [comment]
Actually it seems like the actual blocks are in the `server.rs` handler functions, which are all synchronous.

**Comment by mehmetsefabalik** [comment, marked as answer]
Hey @dessalines , did you find any solution to this?

**Reply by dessalines** [reply]
I think so:
```
/// Handler for Message message.
impl Handler<StandardMessage> for ChatServer {
  type Result = ResponseFuture<Result<String, std::convert::Infallible>>;

  fn handle(&mut self, msg: StandardMessage, ctx: &mut Context<Self>) -> Self::Result {
    let fut = self.parse_json_message(msg, ctx);
    Box::pin(async move {
      match fut.await {
        Ok(m) => {
          // info!("Message Sent: {}", m);
          Ok(m)
        }
        Err(e) => {
          error!("Error during message handling {}", e);
          Ok(e.to_string())
        }
      }
    })
  }
}
```
